shards/shards.go

```diff
@@ -107,6 +111,35 @@ func (ss *shardedSearcher) Close() {
 	}
 }
 
+func selectRepoSet(shards []rankedShard, q query.Q) []rankedShard {
+	filtered := shards[:0]
+
+	eval := query.Map(q, func(q query.Q) query.Q {
+		setQuery, ok := q.(*query.RepoSet)
+		if !ok {
+			return q
+		}
+
+		for _, s := range shards {
+			if repositorer, ok := s.Searcher.(Repositorer); ok {
+				repo := repositorer.Repository()
+				if setQuery.Set[repo.Name] {
+					filtered = append(filtered, s)
```

Comment:
if we have multiple reposets, we could append a shard more than once. My large inline comment below suggests a solution to this.